onViewStateChange callback for viewers

andreasg123 opened the pull request:
This addresses #329. The callback may change the view state, just like the [deck.gl callback](https://deck.gl/docs/api-reference/core/deck#onviewstatechange).

Add "onViewStateChange" to VivViewer, PictureInPictureViewer, SideBySideViewer.
Fix JSDoc `@param` and indicate optional parameters.
Export OverviewLayer.

ilan-gold commented on src/viewers/VivViewer.js:
```diff
@@ -2,12 +2,18 @@ import React, { PureComponent } from 'react'; // eslint-disable-line import/no-u
 import DeckGL from '@deck.gl/react';
 import { getVivId } from '../views/utils';
 
+/**
+ * @callback ViewStateChange
+ * @param {Object} event
+ */
+
```
Can we remove the other two references to this?  It seems like only one is needed for the docs and the docs compiler we use generates multiple references to this.

andreasg123 replied:
I figured out how to import this definition. That's much better than defining it multiple times. Without the import, VS Code doesn't show the correct type on mouse-over.

ilan-gold commented on src/viewers/SideBySideViewer.js:
```diff
@@ -13,14 +18,15 @@ import { SideBySideView, getDefaultInitialViewState } from '../views';
- * @param {number} props.lensBorderColor RGB color of the border of the lens (default [255, 255, 255]).
- * @param {number} props.lensBorderRadius Percentage of the radius of the lens for a border (default 0.02).
- */
+ * @param {boolean} [props.isLensOn] Whether or not to use the lens deafult (false).
```
Didn't know this feature for optional params - thanks!

ilan-gold commented:
@andreasg123 One last thing - can you update the `CHANGELOG.md`.  I need to look into gh actions are not running.

ilan-gold commented:
@andreasg123 Just realized you never requested my review - my apologies if you weren't ready.  I won't review the other PR until you are ready.

andreasg123 commented:
@ilan-gold, I usually don't request reviews when contributing to open source projects, only for internal projects. I'll try to remember to request reviews in the future.

ilan-gold commented:
@andreasg123 Now that I think of it, I suppose I do not either.  Either way then.

ilan-gold commented:
@andreasg123 Is GH actions enabled in your forked repo?  Could you do that so the tests can run?

andreasg123 commented:
I just enabled GH actions in my forked repo. However, I haven't used that before. I see "Node.js CI" but I don't know how to make it run.

ilan-gold commented:
@andreasg123 Can you try making a small change to the Changelog or docs or the like and then pushing? Or try uncommitting, recommitting, and force pushing your last change?

andreasg123 commented:
I figured out why we had different `prettier` results. I installed the current version 2.2.1 globally whereas you selected version 1.19.1 from November 2019. There were some default changes such as `arrowParens`.

Also, `prettier` has a precedence for `package.json` for its configuration. I removed `.prettierrc.json` and added configurations to `package.json`. Now, the only remaining issues are line breaks in a couple of files that I didn't touch.

andreasg123 commented:
@ilan-gold, thanks for making me get familiar with prettier and GH Actions. That will be useful for my own repos.

ilan-gold commented:
> Also, `prettier` has a precedence for `package.json` for its configuration. I removed `.prettierrc.json` and added configurations to `package.json`. Now, the only remaining issues are line breaks in a couple of files that I didn't touch.

Can you specify the issues?  The changes here look fine to me with line breaks since they are >3 arguments.

andreasg123 commented:
`prettier --write` produces this:

```diff
diff --git a/src/layers/ScaleBarLayer.js b/src/layers/ScaleBarLayer.js
index 097894b..9ee0cce 100644
--- a/src/layers/ScaleBarLayer.js
+++ b/src/layers/ScaleBarLayer.js
@@ -122,10 +122,7 @@ export default class ScaleBarLayer extends CompositeLayer {
       coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
       data: [
         {
-          text:
-            String(numUnits)
-              .slice(0, 5)
-              .replace(/\.$/, '') + unit,
+          text: String(numUnits).slice(0, 5).replace(/\.$/, '') + unit,
           position: [xLeftCoord + barLength * 0.5, yCoord + barHeight * 4]
         }
       ],
diff --git a/src/loaders/utils.js b/src/loaders/utils.js
index cc0cdb4..3eacb39 100644
--- a/src/loaders/utils.js
+++ b/src/loaders/utils.js
@@ -149,10 +149,7 @@ export async function getJson(store, key) {
  */
 export function dimensionsFromOMEXML(omexml) {
   const { SizeZ, SizeT, DimensionOrder } = omexml;
-  const dims = DimensionOrder.toLowerCase()
-    .split('')
-    .slice()
-    .reverse();
+  const dims = DimensionOrder.toLowerCase().split('').slice().reverse();
   const dimensions = dims.map(field => {
     if (field === 'x' || field === 'y') {
       return { field, type: 'quantitative', values: null };
```

ilan-gold commented:
@andreasg123 Are you using `npm run format`?  It seems that the branch passes on GH Actions and I cannot reproduce that locally.  `prettier` on the command line uses your globally installed instance I think.

andreasg123 commented:
@ilan-gold, I verified that after uninstalling the global Prettier, `npx prettier --write ...` **does not** produce any changes, including the ones that I listed earlier. Thus, you should be fine as long as you stay with the old version.

ilan-gold commented:
@andreasg123 Ok I will review once the conflicts are resolved and then merge and do a release.

andreasg123 commented:
@ilan-gold, I resolved the conflicts. There were no problems in GH Actions.

ilan-gold commented on src/viewers/VivViewer.js:
```diff
@@ -44,9 +50,19 @@ export default class VivViewer extends PureComponent {
+  _onViewStateChange({ viewId, viewState, interactionState, oldViewState }) {
     // Save the view state and trigger rerender.
-    const { views } = this.props;
+    const { views, onViewStateChange } = this.props;
+    // eslint-disable-next-line no-param-reassign
+    viewState =
+      (onViewStateChange &&
+        onViewStateChange({
+          viewId,
+          viewState,
+          interactionState,
+          oldViewState
+        })) ||
+      viewState;
```
Can you explain the use-case for altering the view state?  I am just noticing this on final review.  I would expect this to simply be a callback since `filterViewState` on the `VivView` class effectively does this.

ilan-gold replied:
I haven't seen hooks like this that alter (some of) their arguments, hence my question.

andreasg123 replied:
I currently don't have a use case for this. At the same time, I see no reason not to expose that part of `Deck.onViewStateChange`: https://deck.gl/docs/api-reference/core/deck#onviewstatechange

ilan-gold replied:
Ok, yes this is what that function does so it's ok to hoist that up to allow for customisation.  Can you add the link to deck.gl docs at the PIP and SBS viewer docs?

andreasg123 replied:
I copied the parameter documentation from `VivViewer` that includes the link.
